### PR TITLE
README rewrite

### PR DESCRIPTION
**Pull request by ezyang**
I rewrote the README. Ignore the Markdown commit.

[Rendered.](https://github.com/ezyang/cabal/blob/156388d14ea9ff7c97a37b38b93f3b2f388d6282/README.md)

### CODE REVIEW

**Comment by phadej** [comment]
So you merged in `HACKING.md` but didn't remove it?

**Comment by ezyang** [comment]
How about that ;)

File: README.md [hunk 1]
```diff
@@ -5,8 +5,148 @@ This Cabal Git repository contains the following packages:
...
+~~~~
+(cd Cabal; cabal install)
+(cd cabal-install; cabal install)
+~~~~
+
+Building Cabal for hacking
+--------------------------
+
+The current recommended way of developing Cabal is to use the
+`new-build` feature which shipped in cabal-install-1.24.  Assuming
```
Comment (23Skidoo) [inline README.md]: I'd add a link to your blog post about nix-local-build here.

**Comment by 23Skidoo** [comment]
LGTM.

File: README.md (hunk 1 above)
Comment (phadej) [inline README.md]: what's the status of `setup-dev.sh` btw?

Comment (ezyang) [inline README.md]: Never head of it. It seems to setup some sort of sandbox? Seems way more complicated then it should be. I'd delete it.

Comment (23Skidoo) [inline README.md]: /cc @BardurArantsson

**Comment by ezyang** [comment]
For what it's worth: how do YOU develop GHC? This document is how I do it, but I don't really know how others go about the workflow.

File: README.md (hunk 1 above)
Comment (BardurArantsson) [inline README.md]: Please don't delete it (yet).

`setup-dev.sh` is the code form of the setup as detailed in `HACKING.md`. It's exactly as complicated as was necessary at the time :).

I haven't tried using the new-build way yet.

**Comment by 23Skidoo** [comment]
I use sandboxes and add-source, but should probably migrate to new-build.

File: README.md (hunk 1 above)
Comment (23Skidoo) [inline README.md]: @BardurArantsson If you haven't noticed, this PR moves all the contents of `HACKING.md` into `README.md` and deletes the former.

**Comment by 23Skidoo** [comment]
Merged, thanks! Leaving the question of whether or not `setup-dev.sh` should be removed open for now.

File: README.md (hunk 1 above)
Comment (BardurArantsson) [inline README.md]: I hadn't noticed, but I don't think it changes anything per se about using `setup-dev.sh`... *until* I can try out new-build and observe that it works for me :).

I've been using `ghcid` to do continuous recompilation, not sure what the status of `ghcid` + new-build is. Once I see that working, I'll be happy to delete `setup-dev.sh` :).

**Pull request merged**